dwave/plugins/torch/autoencoder/ae.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2025 D-Wave
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# The use of the auto encoder implementations below (including the
+# AutoEncoder) with a quantum computing system is
+# protected by the intellectual property rights of D-Wave Quantum Inc.
+# and its affiliates.
+#
+# The use of the auto encoder implementations below (including the
+# AutoEncoder) with D-Wave's quantum computing
+# system will require access to D-Wave’s LeapTM quantum cloud service and
+# will be governed by the Leap Cloud Subscription Agreement available at:
+# https://cloud.dwavesys.com/leap/legal/cloud_subscription_agreement/
+#
+
+from collections.abc import Callable
+
+import torch
+
+__all__ = ["AutoEncoder"]
+
+
+class AutoEncoder(torch.nn.Module):
+    """Autoencoder architecture amenable for training discrete models as priors.
+
+    Such discrete models include spin-variable models amenable for the QPU. This
+    architecture is a modification of the standard autoencoder architecture, where
+    the encoder outputs a latent representation of the data, and the decoder
+    reconstructs the data from the latent representation. In our case, there is an
+    additional step where the latent representation is mapped to a discrete
+    representation, which is then passed to the decoder.
+
+    Args:
+        encoder (torch.nn.Module): The encoder must output latents that are later on
+            passed to `latent_to_discrete`. An encoder has signature (x) -> l. x has
+            shape (batch_size, f1, f2, ...) and l has shape (batch_size, l1, l2, ...).
+        decoder (torch.nn.Module): Decodes discrete tensors into data tensors. A decoder
+            has signature (d) -> x'. d has shape (batch_size, n, d1, d2, ...) and x' has
+            shape (batch_size, f'1, f'2, ...); if x' is the reconstructed data then
+            fi=f'i, but x' might be another representation of the data (e.g. in a
+            text-to-image model, x is a sequence of tokens, and x' is an image). Note
+            that the decoder input is of shape (batch_size, n, d1, d2, ...), where n is
+            a number of discrete representations to be created from a single latent
+            representation of a single initial data point.
+        latent_to_discrete (Callable[[torch.Tensor, int], torch.Tensor] | None): A
+            stochastic function that maps the output of the encoder to a discrete
```

Comment:
> stochastic function

conditional distribution

Reply:
I stick to stochastic function. It is also a conditional distribution.